File: docs/rfc/aleph-008-continuity/prelude/Transpiler.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}

{- | Dhall → Starlark/Nix Transpiler

This is the bridge from typed Dhall BUILD files to:
1. Starlark BUCK files (for Buck2)
2. Nix expressions (for Nix, via dhall-nix or direct)

The key insight: both targets are just data. No complex logic needed.
-}
module Straylight.Transpiler (
    -- * Core transpilation
    toStarlark,
    toNix,
    toBuck,

    -- * Target types
    Target (..),
    RuleKind (..),

    -- * CLI
    main,
) where

import qualified Dhall
import qualified Dhall.Core as D
import qualified Dhall.Map as DM
import qualified Dhall.Nix as DN -- From dhall-nix package

import Data.List (intercalate)
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import System.Environment (getArgs)
import System.FilePath (takeDirectory, (-<.>), (</>))

-- =============================================================================
-- Core Types
-- =============================================================================

-- | A build target extracted from Dhall
data Target = Target
    { targetRule :: RuleKind
    , targetName :: Text
    , targetAttrs :: D.Expr D.Src D.Void
    }
    deriving (Show)

-- | Rule kinds we support
data RuleKind
    = RustLibrary
    | RustBinary
    | RustTest
    | LeanLibrary
    | LeanBinary
    | LeanTest
    | ProvenLibrary
    | CxxLibrary
    | CxxBinary
    | CxxTest
    | HaskellLibrary
    | HaskellBinary
    | HaskellTest
    | PureScriptLibrary
    | PureScriptBundle
    | PureScriptTest
    | NvLibrary
    | NvBinary
    | NvTest
    | Alias
    | Filegroup
    deriving (Show, Eq)

-- =============================================================================
-- Dhall → Starlark
-- =============================================================================

-- | Convert a Dhall expression to Starlark syntax
toStarlark :: D.Expr s a -> Text
toStarlark = \case
    -- Records become dicts or kwargs
    D.RecordLit fields ->
        let pairs =
                map
                    (\(k, v) -> k <> " = " <> toStarlark (D.recordFieldValue v))
                    (DM.toList fields)
         in "{" <> T.intercalate ", " pairs <> "}"
    -- Lists become lists
    D.ListLit _ items ->
        let elems = map toStarlark (toList items)
         in "[" <> T.intercalate ", " elems <> "]"
    -- Text becomes strings
    D.TextLit chunks ->
        "\"" <> escapeStarlark (textFromChunks chunks) <> "\""
    -- Booleans
    D.BoolLit True -> "True"
    D.BoolLit False -> "False"
    -- Numbers
    D.NaturalLit n -> T.pack (show n)
    D.IntegerLit n -> T.pack (show n)
    D.DoubleLit n -> T.pack (show (D.denote n))
    -- Optional
    D.Some inner -> toStarlark inner
    D.App D.None _ -> "None"
    -- Unions (enums) - just use the tag as a string
    D.Field (D.Union _) (D.FieldSelection _ name _) ->
        "\"" <> name <> "\""
    -- Union application (variant with value)
    D.App (D.Field (D.Union _) (D.FieldSelection _ name _)) value ->
        -- For complex variants, emit as struct
        name <> "(" <> toStarlark value <> ")"
    -- Variables shouldn't appear in normalized output
    D.Var _ -> error "Unexpected variable in normalized Dhall"
    -- Fallback
    other -> "# TODO: " <> T.pack (show other)
  where
    toList = foldr (:) []

-- | Escape a string for Starlark
escapeStarlark :: Text -> Text
escapeStarlark = T.concatMap $ \case
    '\\' -> "\\\\"
    '"' -> "\\\""
    '\n' -> "\\n"
    '\r' -> "\\r"
    '\t' -> "\\t"
    c -> T.singleton c

-- | Extract text from Dhall text literal chunks
textFromChunks :: D.Chunks s a -> Text
textFromChunks (D.Chunks [] t) = t
textFromChunks (D.Chunks ((t, _) : rest) final) =
    t <> textFromChunks (D.Chunks rest final)

-- =============================================================================
-- Dhall → Nix
-- =============================================================================

{- | Convert a Dhall expression to Nix syntax
Uses dhall-nix for the heavy lifting
-}
toNix :: D.Expr D.Src D.Void -> Text
toNix expr = DN.dhallToNix expr -- dhall-nix does the work

-- =============================================================================
-- Emit BUCK file
-- =============================================================================

-- | Emit a complete BUCK file from a list of targets
toBuck :: [Target] -> Text
toBuck targets =
    T.unlines $
        [ "# Generated from BUILD.dhall - DO NOT EDIT"
        , "# Regenerate with: straylight gen"
        , ""
        ]
            ++ concatMap emitTarget targets

-- | Emit a single target
emitTarget :: Target -> [Text]
emitTarget Target{..} =
    [ ruleName targetRule <> "("
    , "    name = \"" <> targetName <> "\","
    ]
        ++ attrLines
        ++ [")"]
  where
    attrLines = case targetAttrs of
        D.RecordLit fields ->
            [ "    " <> k <> " = " <> toStarlark (D.recordFieldValue v) <> ","
            | (k, v) <- DM.toList fields
            , k /= "common" -- Skip common, we extracted name from it
            ]
        _ -> []

-- | Rule name in Starlark
ruleName :: RuleKind -> Text
ruleName = \case
    RustLibrary -> "rust_library"
    RustBinary -> "rust_binary"
    RustTest -> "rust_test"
    LeanLibrary -> "lean_library"
    LeanBinary -> "lean_binary"
    LeanTest -> "lean_test"
    ProvenLibrary -> "proven_library"
    CxxLibrary -> "cxx_library"
    CxxBinary -> "cxx_binary"
    CxxTest -> "cxx_test"
    HaskellLibrary -> "haskell_library"
    HaskellBinary -> "haskell_binary"
    HaskellTest -> "haskell_test"
    PureScriptLibrary -> "purescript_library"
    PureScriptBundle -> "purescript_bundle"
    PureScriptTest -> "purescript_test"
    NvLibrary -> "nv_library"
    NvBinary -> "nv_binary"
    NvTest -> "nv_test"
    Alias -> "alias"
    Filegroup -> "filegroup"

-- =============================================================================
-- Parse targets from Dhall
-- =============================================================================

{- | Extract targets from a normalized Dhall expression
Expects: List { rule : Text, attrs : {...} }
-}
extractTargets :: D.Expr D.Src D.Void -> Either Text [Target]
extractTargets (D.ListLit _ items) = traverse extractTarget (toList items)
  where
    toList = foldr (:) []
extractTargets other = Left $ "Expected list of targets, got: " <> T.pack (show other)

extractTarget :: D.Expr D.Src D.Void -> Either Text Target
extractTarget (D.RecordLit fields) = do
    rule <- case DM.lookup "rule" fields of
        Just (D.RecordField _ (D.TextLit (D.Chunks [] r)) _ _) -> parseRule r
        _ -> Left "Missing or invalid 'rule' field"

    attrs <- case DM.lookup "attrs" fields of
        Just (D.RecordField _ a _ _) -> Right a
        _ -> Left "Missing 'attrs' field"

    name <- case attrs of
        D.RecordLit fs -> case DM.lookup "common" fs of
            Just (D.RecordField _ (D.RecordLit common) _ _) ->
                case DM.lookup "name" common of
                    Just (D.RecordField _ (D.TextLit (D.Chunks [] n)) _ _) -> Right n
                    _ -> Left "Missing name in common attrs"
            _ -> Left "Missing common attrs"
        _ -> Left "attrs is not a record"

    return Target{targetRule = rule, targetName = name, targetAttrs = attrs}
extractTarget other = Left $ "Expected record, got: " <> T.pack (show other)

parseRule :: Text -> Either Text RuleKind
parseRule = \case
    "rust_library" -> Right RustLibrary
    "rust_binary" -> Right RustBinary
    "rust_test" -> Right RustTest
    "lean_library" -> Right LeanLibrary
    "lean_binary" -> Right LeanBinary
    "lean_test" -> Right LeanTest
    "proven_library" -> Right ProvenLibrary
    "cxx_library" -> Right CxxLibrary
    "cxx_binary" -> Right CxxBinary
    "cxx_test" -> Right CxxTest
    "haskell_library" -> Right HaskellLibrary
    "haskell_binary" -> Right HaskellBinary
    "haskell_test" -> Right HaskellTest
    "purescript_library" -> Right PureScriptLibrary
    "purescript_bundle" -> Right PureScriptBundle
    "purescript_test" -> Right PureScriptTest
    "nv_library" -> Right NvLibrary
    "nv_binary" -> Right NvBinary
    "nv_test" -> Right NvTest
    "alias" -> Right Alias
    "filegroup" -> Right Filegroup
    other -> Left $ "Unknown rule: " <> other

-- =============================================================================
-- CLI
-- =============================================================================

main :: IO ()
main = do
    args <- getArgs
    case args of
        ["--help"] -> putStrLn usage
        [input] -> transpileFile input "buck"
        [input, "--nix"] -> transpileFile input "nix"
        [input, "--buck"] -> transpileFile input "buck"
        _ -> putStrLn usage

usage :: String
usage =
    unlines
        [ "straylight-transpile: Dhall → Starlark/Nix"
        , ""
        , "Usage: straylight-transpile BUILD.dhall [--buck|--nix]"
        , ""
        , "  --buck  Output BUCK file (default)"
        , "  --nix   Output Nix expression"
        ]

transpileFile :: FilePath -> String -> IO ()
transpileFile input format = do
    -- Parse and normalize Dhall
    expr <- Dhall.inputExpr (T.pack input)

    case extractTargets expr of
        Left err -> error $ T.unpack err
        Right targets -> do
            let output = case format of
                    "nix" -> T.unlines $ map (toNix . targetAttrs) targets
                    "buck" -> toBuck targets
                    _ -> toBuck targets

            let outFile = case format of
                    "nix" -> input -<.> "nix"
                    "buck" -> takeDirectory input </> "BUCK"
                    _ -> takeDirectory input </> "BUCK"

            TIO.writeFile outFile output
            putStrLn $ "Wrote " ++ outFile

File: docs/rfc/aleph-008-continuity/dhall/transpiler.rs
//! Dhall → Starlark Transpiler
//!
//! This is a sketch of the transpiler. It's ~150 lines to handle the core cases.
//! The key insight: BUCK files are just data (function calls with kwargs),
//! so transpilation is mechanical.

use dhall::semantics::Nir;
use dhall::syntax::Expr;

/// Convert a Dhall value to Starlark syntax
pub fn to_starlark(value: &Nir) -> String {
    use dhall::semantics::NirKind::*;
    
    match value.kind() {
        // Records become keyword arguments or dicts
        RecordLit(fields) => {
            let pairs: Vec<String> = fields
                .iter()
                .map(|(k, v)| format!("{} = {}", k, to_starlark(v)))
                .collect();
            format!("{{{}}}", pairs.join(", "))
        }
        
        // Lists become lists
        List(items) => {
            let elements: Vec<String> = items
                .iter()
                .map(|v| to_starlark(v))
                .collect();
            format!("[{}]", elements.join(", "))
        }
        
        // Text becomes strings
        TextLit(chunks) => {
            // Handle interpolation if needed
            let s = chunks.to_string();
            format!("\"{}\"", escape_starlark(&s))
        }
        
        // Booleans
        BoolLit(true) => "True".to_string(),
        BoolLit(false) => "False".to_string(),
        
        // Numbers
        NaturalLit(n) => n.to_string(),
        IntegerLit(n) => n.to_string(),
        DoubleLit(n) => format!("{}", n),
        
        // None/Some for Optional
        EmptyOptionalLit(_) => "None".to_string(),
        NEOptionalLit(inner) => to_starlark(inner),
        
        // Unions (enums) - depends on schema
        // For now, just use the variant name as a string
        UnionConstructor(name, _) => format!("\"{}\"", name),
        UnionLit(name, inner, _) => {
            // Could be struct-like or just a tag
            if inner.kind().is_record_lit() {
                format!("{}({})", name, to_starlark(inner))
            } else {
                format!("\"{}\"", name)
            }
        }
        
        // Variables shouldn't appear in normalized output
        Var(_) => panic!("Unexpected variable in normalized Dhall"),
        
        // Everything else
        _ => format!("# TODO: {:?}", value.kind()),
    }
}

fn escape_starlark(s: &str) -> String {
    s.replace('\\', "\\\\")
        .replace('"', "\\\"")
        .replace('\n', "\\n")
        .replace('\r', "\\r")
        .replace('\t', "\\t")
}

/// Emit a Buck2 target from a Dhall rule invocation
pub fn emit_target(rule_name: &str, attrs: &Nir) -> String {
    use dhall::semantics::NirKind::*;
    
    match attrs.kind() {
        RecordLit(fields) => {
            let kwargs: Vec<String> = fields
                .iter()
                .map(|(k, v)| format!("    {} = {},", k, to_starlark(v)))
                .collect();
            format!("{}(\n{}\n)", rule_name, kwargs.join("\n"))
        }
        _ => panic!("Target attributes must be a record"),
    }
}

/// Convert a list of targets to a BUCK file
pub fn emit_buck_file(targets: &[(String, Nir)]) -> String {
    let mut lines = vec![
        "# Generated from BUILD.dhall - DO NOT EDIT".to_string(),
        "# Regenerate with: straylight gen".to_string(),
        "".to_string(),
    ];
    
    for (rule_name, attrs) in targets {
        lines.push(emit_target(rule_name, attrs));
        lines.push("".to_string());
    }
    
    lines.join("\n")
}

// -----------------------------------------------------------------------------
// Example usage (pseudocode)
// -----------------------------------------------------------------------------

/*
fn main() -> Result<()> {
    // 1. Parse and evaluate Dhall
    let cx = Ctxt::new();
    let parsed = Parsed::parse_file(Path::new("BUILD.dhall"))?;
    let resolved = parsed.resolve(cx)?;
    let typed = resolved.typecheck(cx)?;
    let normalized = typed.normalize(cx);
    
    // 2. Extract targets from the normalized value
    //    Assumes BUILD.dhall evaluates to a list of { rule : Text, attrs : {...} }
    let targets = extract_targets(normalized.as_nir())?;
    
    // 3. Emit BUCK file
    let buck_content = emit_buck_file(&targets);
    
    // 4. Write to BUCK
    std::fs::write("BUCK", buck_content)?;
    
    Ok(())
}
*/

// -----------------------------------------------------------------------------
// What BUILD.dhall looks like
// -----------------------------------------------------------------------------

/*
-- BUILD.dhall
let DICE = https://straylight.cx/dice/v1/package.dhall sha256:...

let mylib = DICE.rust_library
    { name = "mylib"
    , srcs = [ "src/lib.rs", "src/foo.rs" ]
    , deps = [ "//other:lib" ]
    , edition = DICE.RustEdition.Edition2024
    }

let mybin = DICE.rust_binary
    { name = "mybin"
    , srcs = [ "src/main.rs" ]
    , deps = [ ":mylib" ]
    }

in  [ { rule = "rust_library", attrs = mylib }
    , { rule = "rust_binary", attrs = mybin }
    ]
*/

// -----------------------------------------------------------------------------
// Generated BUCK output
// -----------------------------------------------------------------------------

/*
# Generated from BUILD.dhall - DO NOT EDIT
# Regenerate with: straylight gen

rust_library(
    name = "mylib",
    srcs = ["src/lib.rs", "src/foo.rs"],
    deps = ["//other:lib"],
    edition = "2024",
)

rust_binary(
    name = "mybin",
    srcs = ["src/main.rs"],
    deps = [":mylib"],
)
*/
